R/selVarGenes.R

```diff
@@ -0,0 +1,405 @@
+#' @param loessModel the loess fit to all points.
+#' @param x the x values of all points.
+#' @param y the y values of all points.
+#' @param x_curve the x values of the points on the loess curve.
+#' @param method distance measure to be used.
+#' @param ... additional parameters for the \code{dist.matrix} function.
+#' 
+#' @return An object of class \code{matrix} and \code{dist.matrix}, containing the
+#'    distances between all points to the specified points on the loess curve.
+#'
+#' @examples 
+#'   library(stats)
+#'   
+#'   # create data
+#'   x <- 1:1000
+#'   y <- sort(rnorm(n = 1000, mean = 0, sd = 5)) + runif(n = 1000, min = 0, max = 10)
+#'   lo <- loess(y ~ x)
+#'   x_curve <- seq(range(x)[1], range(x)[2], length.out=100)
+#'   
+#'   # plot
+#'   plot(x, y, pch = 16, col="#BEBEBE90")
+#'   lines(x_curve, predict(lo, newdata=x_curve), col="red")
+#'   
+#'   # get euclidean distance of all points to the curve (red)
+#'   euclMat <- getDistMat(loessModel=lo, x=x, y=y, x_curve=x_curve)
+#'   euclMat[1:6, 1:6]
+#'
+#' @importFrom wordspace dist.matrix
+#' @importFrom stats predict
+#' 
+#' @export
+#' 
+getDistMat <- function(loessModel = NULL, x = NULL, y = NULL, x_curve = NULL, method = "euclidean", ...){
+     
+     ## checks
+     stopifnot(!is.null(x_curve))
+     stopifnot(!is.null(x))
+     stopifnot(!is.null(y))
+     stopifnot(!is.null(loessModel))
+     stopifnot(length(x) == length(y))
+     stopifnot(is.numeric(x_curve))
+     stopifnot(is.numeric(x))
+     stopifnot(is.numeric(y))
+
+     ## predict y_curv for x_curve using loessModel
+     y_curve <- stats::predict(loessModel, newdata = x_curve)
+     
+     ## get distances
+     distMat <- wordspace::dist.matrix(M = cbind(x, y), M2 = cbind(x_curve, y_curve), method = method, ... )
+     colnames(distMat) <- paste0("pointOnCurve", seq_len(length(y_curve)))
+     if (!is.null(names(x))) {
+          rownames(distMat) <- names(x)
+     } else {
+          rownames(distMat) <- paste0("xy", seq_len(length(x)))
+     }
+     
+     ## return
+     distMat
+     
+}
+
+
+#' @title Select Variable Genes in Single Cell RNA-seq
+#' 
+#' @description This function selects the most variable genes from a \code{SingleCellExperiment} object
+#'   using the plot that displays the log2 coefficient of variation as a function of the log2 mean for 
+#'   all genes across all the cells.
+#' 
+#' @author Dania Machlab 
+#' 
+#' @param sce \code{SingleCellExperiment} object containing the raw counts and the size factors.
+#' @param Nmads number of MADs beyond which genes are selected per bin.
+#' @param minCells keep genes with minimum expression in at least this number of cells.
+#' @param minExpr keep genes with expression greater than or equal to this in \code{minCells} cells using 
+#'    the raw count matrix corrected for library size differences using the sizeFactors from \code{sce}.
+#' @param topExprPerc the top percent expressed genes that will be excluded from the loess fit. 
```

Comment:
topExprFrac instead of topExprPerc? As far as I can see, it should be a value between 0 and 1. With "Top percent", it's not completely clear if it could be e.g. 20 (percent).